storage/paths/index.go

```diff
@@ -415,6 +907,193 @@ func (i *Index) StorageDropSector(ctx context.Context, storageID storiface.ID, s
+		Can_store   bool
+		Groups      string
+		Allow_to    string
+		Allow_types string
+		Deny_types  string
+	}
+
+	var rows []dbRes
+
+	fts := ft.AllSet()
+	// Find all storage info which already hold this sector + filetype
+	err := dbi.harmonyDB.Select(ctx, &rows,
+		`		SELECT DISTINCT ON (stor.storage_id)
+						  stor.storage_id,
+						  COUNT(*) OVER(PARTITION BY stor.storage_id) as count, 
+						  BOOL_OR(is_primary) OVER(PARTITION BY stor.storage_id) AS is_primary,
+						  urls,
+						  weight,
+						  can_seal,
+						  can_store,
+						  groups,
+						  allow_to,
+						  allow_types,
+						  deny_types
+						FROM sectorlocation sec
+						JOIN storagelocation stor ON sec.storage_id = stor.storage_id 
+						WHERE sec.miner_id = $1
+						  AND sec.sector_num = $2 
+						  AND sec.sector_filetype = ANY($3)  
+						ORDER BY stor.storage_id`,
+		s.Miner, s.Number, fts)
+	if err != nil {
+		return nil, xerrors.Errorf("Finding sector storage from DB fails with err: ", err)
+	}
+
+	for _, row := range rows {
+
+		// Parse all urls
+		var urls, burls []string
+		for _, u := range splitString(row.Urls) {
+			rl, err := url.Parse(u)
+			if err != nil {
+				return nil, xerrors.Errorf("failed to parse url: %w", err)
+			}
+			rl.Path = gopath.Join(rl.Path, ft.String(), storiface.SectorName(s))
+			urls = append(urls, rl.String())
+			burls = append(burls, u)
+		}
+
+		result = append(result, storiface.SectorStorageInfo{
+			ID:         storiface.ID(row.Storage_id),
+			URLs:       urls,
+			BaseURLs:   burls,
+			Weight:     row.Weight * row.Count,
+			CanSeal:    row.Can_seal,
+			CanStore:   row.Can_store,
+			Primary:    row.Is_primary,
+			AllowTypes: splitString(row.Allow_types),
+			DenyTypes:  splitString(row.Deny_types),
+		})
+
+		storageWithSector[row.Storage_id] = true
+
+		allowTo := splitString(row.Allow_to)
+		if allowList != nil && len(allowTo) > 0 {
+			for _, group := range allowTo {
+				allowList[group] = struct{}{}
+			}
+		} else {
+			allowList = nil // allow to any
+		}
+	}
+
+	// Find all storage paths which can hold this sector if allowFetch is true
+	if allowFetch {
+		spaceReq, err := ft.SealSpaceUse(ssize)
+		if err != nil {
+			return nil, xerrors.Errorf("estimating required space: %w", err)
+		}
+
+		// Conditions to satisfy when choosing a sector
+		// 1. CanSeal is true
+		// 2. Available >= spaceReq
+		// 3. curr_time - last_heartbeat < SkippedHeartbeatThresh
+		// 4. heartbeat_err is NULL
+		// 5. not one of the earlier picked storage ids
+		// 6. !ft.AnyAllowed(st.info.AllowTypes, st.info.DenyTypes)
+		// 7. Storage path is part of the groups which are allowed from the storage paths which already hold the sector
+
+		var rows []struct {
+			Storage_id  string
+			Urls        string
+			Weight      uint64
+			Can_seal    bool
+			Can_store   bool
+			Groups      string
+			Allow_types string
+			Deny_types  string
+		}
+		err = dbi.harmonyDB.Select(ctx, &rows,
+			`select storage_id, 
+					  	urls, 
```

Comment:
please use spaces instead of tabs for this as it formats weird for different editors. Also seen above.

Reply:
Do you have a preferred whitespace formatter we can run on the code?
Ideally, this would be done as part of make gen